File: tests/test_market_sim.py
import pytest

try:
    from trading_engine import AssetConfig, Engine, MultiAssetMarketSim, RiskConfig
except ImportError:
    MultiAssetMarketSim = None


@pytest.fixture
def sim_setup():
    if MultiAssetMarketSim is None:
        pytest.skip("trading_engine extension is not built")

    risk = RiskConfig(
        max_order_qty=1000.0,
        max_order_notional=10_000_000.0,
        max_position_notional=20_000_000.0,
        price_collar_pct=0.25,
        max_drawdown_pct=0.30,
        max_orders_per_sec=100_000,
        require_margin=False,
    )

    engine = Engine(initial_balance=500_000.0, leverage=2.0, risk_config=risk)

    btc = AssetConfig("BTC-USDT", initial_price=60_000.0, drift=0.05, volatility=0.40)
    eth = AssetConfig("ETH-USDT", initial_price=3_200.0, drift=0.05, volatility=0.50)
    sol = AssetConfig("SOL-USDT", initial_price=150.0, drift=0.08, volatility=0.65)

    corr = [
        [1.0, 0.85, 0.70],
        [0.85, 1.0, 0.75],
        [0.70, 0.75, 1.0],
    ]

    sim = MultiAssetMarketSim(engine, [btc, eth, sol], corr, seed=123)
    return engine, sim


def test_multi_asset_sim_stepping(sim_setup):
    engine, sim = sim_setup

    prices = sim.step(dt=1.0)
    assert len(prices) == 3
    assert "BTC-USDT" in prices
    assert "ETH-USDT" in prices
    assert "SOL-USDT" in prices

    sim.run_steps(count=5, dt=1.0)
    assert sim.step_count() == 6

    # Verify order books are active and populated
    for sym in ["BTC-USDT", "ETH-USDT", "SOL-USDT"]:
        depth = engine.get_depth(sym, levels=3)
        assert depth is not None
        assert depth.best_bid() is not None
        assert depth.best_ask() is not None
        assert depth.best_bid() < depth.best_ask()
        assert depth.spread() > 0.0


def test_strategy_trading_against_simulated_market(sim_setup):
    engine, sim = sim_setup

    # Run initial steps to create market liquidity
    sim.run_steps(count=3, dt=1.0)

    initial_pos = engine.get_position("BTC-USDT")
    initial_qty = initial_pos.quantity if initial_pos else 0.0

    # Strategy buys 0.1 BTC using a market order
    order = engine.submit_order(
        symbol="BTC-USDT",
        side="BUY",
        order_type="MARKET",
        price=0.0,
        quantity=0.1,
        time_in_force="IOC",
    )
    assert order.filled_quantity == 0.1

    pos = engine.get_position("BTC-USDT")
    assert pos is not None
    assert pos.quantity == pytest.approx(initial_qty + 0.1)

File: tests/test_market_sim.rs
use std::sync::Arc;
use trading_engine::market_sim::{cholesky, AssetConfig, MultiAssetMarketSim};
use trading_engine::{ExecutionEngine, RiskConfig};

#[test]
fn test_cholesky_decomposition_2x2() {
    let corr = vec![vec![1.0, 0.6], vec![0.6, 1.0]];
    let l = cholesky(&corr).unwrap();

    // L * L^T should equal corr
    let l00 = l[0][0];
    let l10 = l[1][0];
    let l11 = l[1][1];

    assert!((l00 * l00 - 1.0).abs() < 1e-6);
    assert!((l10 * l00 - 0.6).abs() < 1e-6);
    assert!((l10 * l10 + l11 * l11 - 1.0).abs() < 1e-6);
}

#[test]
fn test_cholesky_decomposition_3x3() {
    let corr = vec![
        vec![1.0, 0.8, 0.5],
        vec![0.8, 1.0, 0.4],
        vec![0.5, 0.4, 1.0],
    ];
    let l = cholesky(&corr).unwrap();

    // Reconstruct and verify
    for i in 0..3 {
        for j in 0..3 {
            let mut sum = 0.0;
            for k in 0..3 {
                sum += l[i][k] * l[j][k];
            }
            assert!((sum - corr[i][j]).abs() < 1e-5);
        }
    }
}

#[test]
fn test_multi_asset_market_sim_flow() {
    let risk_config = RiskConfig {
        max_order_qty: 1000.0,
        max_order_notional: 10_000_000.0,
        max_position_notional: 20_000_000.0,
        price_collar_pct: 0.20,
        max_drawdown_pct: 0.30,
        max_orders_per_sec: 100_000,
        require_margin: false,
    };

    let engine = Arc::new(ExecutionEngine::new(1_000_000.0, 1.0, risk_config));

    let assets = vec![
        AssetConfig::new("BTC-USDT", 50_000.0, 0.05, 0.40, 0.50, 0.001),
        AssetConfig::new("ETH-USDT", 3_000.0, 0.05, 0.50, 0.10, 0.01),
    ];

    let corr = vec![vec![1.0, 0.85], vec![0.85, 1.0]];

    let mut sim = MultiAssetMarketSim::new(engine.clone(), assets, corr, 12345).unwrap();

    // Step the simulation 5 times
    for _ in 0..5 {
        sim.step(1.0);
    }

    assert_eq!(sim.step_count(), 5);

    // Verify order books are seeded with active quotes
    let btc_depth = engine.get_depth("BTC-USDT", 5).unwrap();
    let eth_depth = engine.get_depth("ETH-USDT", 5).unwrap();

    assert!(btc_depth.best_bid().is_some());
    assert!(btc_depth.best_ask().is_some());
    assert!(btc_depth.best_bid().unwrap() < btc_depth.best_ask().unwrap());

    assert!(eth_depth.best_bid().is_some());
    assert!(eth_depth.best_ask().is_some());
    assert!(eth_depth.best_bid().unwrap() < eth_depth.best_ask().unwrap());

    let prices = sim.get_prices();
    assert!(prices.contains_key("BTC-USDT"));
    assert!(prices.contains_key("ETH-USDT"));
}
